src/librustc_typeck/check/coercion.rs

```diff
@@ -424,20 +425,32 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
             debug!("Missing Unsize or CoerceUnsized traits");
             return Err(TypeError::Mismatch);
         };
+        let mut leftover_predicates = vec![];
 
-        // Note, we want to avoid unnecessary unsizing. We don't want to coerce to
-        // a DST unless we have to. This currently comes out in the wash since
-        // we can't unify [T] with U. But to properly support DST, we need to allow
-        // that, at which point we will need extra checks on the target here.
+        // Subtype the whole source type first, allowing, e.g.
+        // coercing `&'static [&'static T; N]` to `&'b [&'a T]`,
+        // or `Rc<[&'static T; N]>` to `Rc<[&'a T]>`, etc.
+        let coercion = TypeVariableOrigin::MiscVariable(self.cause.span);
+        let super_source = self.next_ty_var(coercion);
+        match self.sub_types(false, &self.cause, source, super_source)? {
```

Comment:
Not needing `use_lub` felt wrong, and it is!
Reordering [these two](https://github.com/rustless/valico/blob/ba11d71b6f5210d0ea1740828f2a54881f2eff74/src/json_dsl/param.rs#L136-L140) `match` arms "fixes" the false positive, but they should both work identically.
I'm preparing an improved version that should not have this problem.